### PR TITLE
1d

### PR DESCRIPTION
**Pull request by 13bscsaamjad**
added oned ctrl for multiple points per step in a step scan.

### CODE REVIEW

File: sardana_albaem/ctrl/Albaem2OneDCtrl.py [hunk 1]
```diff
@@ -0,0 +1,486 @@
...
+from sardana.pool import AcqSynch
+from sardana.pool.controller import OneDController, Type, Access, \
+    Description, Memorize, Memorized, NotMemorized, FGet, FSet, DefaultValue
+from sardana.sardanavalue import SardanaValue
+from functools import wraps, partial
+import six
+
+__all__ = ['Albaem2OneDCtrl']
+
+TRIGGER_INPUTS = {'DIO_1': 0, 'DIO_2': 1, 'DIO_3': 2, 'DIO_4': 3,
```
Comment (meguiraun) [inline sardana_albaem/ctrl/Albaem2OneDCtrl.py]: where is this used?

File: sardana_albaem/ctrl/Albaem2OneDCtrl.py [hunk 2]
```diff
@@ -0,0 +1,486 @@
...
+    @handle_error(msg="__init__: Could not connect to the device!")
+    def __init__(self, inst, props, *args, **kwargs):
+        """Class initialization."""
+        OneDController.__init__(self, inst, props, *args, **kwargs)
+        self.ip_config = (self.AlbaEmHost, self.Port)
+        self.albaem_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+        self.albaem_socket.settimeout(1)
+        self.albaem_socket.connect(self.ip_config)
+        self.itime = 0.0
+        self.master = None
```
Comment (meguiraun) [inline sardana_albaem/ctrl/Albaem2OneDCtrl.py:124]: what is this for?

File: sardana_albaem/ctrl/Albaem2OneDCtrl.py [hunk 3]
```diff
@@ -0,0 +1,486 @@
...
+        self._latency_time = 0.001  # In fact, it is just 320us
```
Comment (meguiraun) [inline sardana_albaem/ctrl/Albaem2OneDCtrl.py:125]: and this?

File: sardana_albaem/ctrl/Albaem2OneDCtrl.py [hunk 4]
```diff
@@ -0,0 +1,486 @@
...
+        if axis != 1:
+            raise Exception('The master channel should be the axis 1')
+
+        self.itime = value
+
+        # Set Integration time in ms
+        val = self.itime * 1000
+        if val < 0.1:   # minimum integration time 
+            self._log.debug("The minimum integration time is 0.1 ms")
+            val = 0.1
```
Comment (meguiraun) [inline sardana_albaem/ctrl/Albaem2OneDCtrl.py]: is there any reason for not raising an exception instead? You are correcting the user silently

File: sardana_albaem/ctrl/Albaem2OneDCtrl.py [hunk 5]
```diff
@@ -0,0 +1,486 @@
...
+        """Class initialization."""
+        OneDController.__init__(self, inst, props, *args, **kwargs)
+        self.ip_config = (self.AlbaEmHost, self.Port)
+        self.albaem_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+        self.albaem_socket.settimeout(1)
+        self.albaem_socket.connect(self.ip_config)
+        self.itime = 0.0
+        self.master = None
+        self._latency_time = 0.001  # In fact, it is just 320us
+        self._repetitions = 0
```
Comment (meguiraun) [inline sardana_albaem/ctrl/Albaem2OneDCtrl.py:126]: what is repetitions for? and its relationship with pointsPerStep?

File: sardana_albaem/ctrl/Albaem2OneDCtrl.py [hunk 6]
```diff
@@ -0,0 +1,486 @@
...
+                self._repetitions = self._points_per_step
+        self.sendCmd('TRIG:MODE %s' % source)
+        if self._synchronization in [AcqSynch.HardwareTrigger,
+                                     AcqSynch.HardwareGate]:
+            self.sendCmd('TRIG:INPU %s' % self.ExtTriggerInput)
+        # Set Number of Triggers
+        self.sendCmd('ACQU:NTRI %r' % self._repetitions)
+        
+        # Array of arrays for ID readings from all channels
+        self.new_data = []
```
Comment (AntoineDupre) [inline sardana_albaem/ctrl/Albaem2OneDCtrl.py]: just passing by this (you know nostalgia ...) and i can not resist to point out that this line looks pointless :trollface:

File: sardana_albaem/ctrl/Albaem2OneDCtrl.py (hunk 5 above)
Comment (meguiraun) [reply sardana_albaem/ctrl/Albaem2OneDCtrl.py:126]: should not be called NumberOfTriggers?

File: sardana_albaem/ctrl/Albaem2OneDCtrl.py [hunk 7]
```diff
@@ -0,0 +1,486 @@
...
+                                     AcqSynch.SoftwareGate]:
+            # The HW needs the software trigger
+            # APPEND SWTRIG TO THE START COMMAND OR SEND ANOTHER COMMAND
+            # TRIG:SWSEt
+            cmd += ' SWTRIG'
+
+        self.sendCmd(cmd)
+        # THIS PROTECTION HAS TO BE REVIEWED
+        # FAST INTEGRATION TIMES MAY RAISE WRONG EXCEPTIONS
+        # e.g. 10ms ACQTIME -> self.state MAY BE NOT MOVING BECAUSE
```
Comment (meguiraun) [inline sardana_albaem/ctrl/Albaem2OneDCtrl.py:244]: a small sleep would help, horrible I know...

File: sardana_albaem/ctrl/Albaem2OneDCtrl.py [hunk 8]
```diff
@@ -0,0 +1,486 @@
...
+        return True
+
+    @debug_it
+    def StartOne(self, axis, value):
+        pass
+
+    @debug_it
+    @handle_error(msg="ReadAll: Unable to read from the device!")
+    def ReadAll(self):
+        self.new_data = []
```
Comment (meguiraun) [inline sardana_albaem/ctrl/Albaem2OneDCtrl.py]: unneeded line

File: sardana_albaem/ctrl/Albaem2OneDCtrl.py [hunk 9]
```diff
@@ -0,0 +1,486 @@
...
+            # Apply the formula for each value
+            formula = self.formulas[axis]
+            formula = formula.lower()
+            values_formula = [eval(formula, {'value': val}) for val
+                                in values]
+            self.new_data[axis].extend(values_formula)
+            axis +=1
+        time_data = [self.itime] * len(self.new_data[1])
+        self.new_data[0] = (time_data)
+
```
Comment (meguiraun) [inline sardana_albaem/ctrl/Albaem2OneDCtrl.py:286]: missing `return self.new_data`?

File: sardana_albaem/ctrl/Albaem2OneDCtrl.py [hunk 10]
```diff
@@ -0,0 +1,486 @@
...
+    @debug_it
+    @handle_error(msg="AbortOne: Could not abort device!")
+    def AbortOne(self, axis):
+        self.sendCmd('ACQU:STOP')
+        self._is_aborted = True
+
+    @debug_it
+    @handle_error(msg="sendCmd: Could not configure device!")
+    def sendCmd(self, cmd, rw=True, size=8096):
```
Comment (meguiraun) [inline sardana_albaem/ctrl/Albaem2OneDCtrl.py:308]: minor, but can we move this out of this file and import it?

File: sardana_albaem/ctrl/Albaem2OneDCtrl.py [hunk 11]
```diff
@@ -0,0 +1,486 @@
...
+    }
+
+    @handle_error(msg="__init__: Could not connect to the device!")
+    def __init__(self, inst, props, *args, **kwargs):
+        """Class initialization."""
+        OneDController.__init__(self, inst, props, *args, **kwargs)
+        self.ip_config = (self.AlbaEmHost, self.Port)
+        self.albaem_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+        self.albaem_socket.settimeout(1)
+        self.albaem_socket.connect(self.ip_config)
```
Comment (meguiraun) [inline sardana_albaem/ctrl/Albaem2OneDCtrl.py:122]: this out of this file together with the `sendCmd?`

File: sardana_albaem/ctrl/Albaem2OneDCtrl.py [hunk 12]
```diff
@@ -0,0 +1,486 @@
...
+
+    @debug_it
+    @handle_error(msg="get_PointsPerStep:")
+    def get_PointsPerStep(self):
+        return self._points_per_step
+
+    @debug_it
+    @handle_error(msg="set_PointsPerStep:")
+    def set_PointsPerStep(self, value):
+        self._points_per_step = value
```
Comment (meguiraun) [inline sardana_albaem/ctrl/Albaem2OneDCtrl.py]: how one is supposed to update this value before an scan?

File: sardana_albaem/ctrl/Albaem2OneDCtrl.py (hunk 6 above)
Comment (meguiraun) [reply sardana_albaem/ctrl/Albaem2OneDCtrl.py]: hahahhahhahaha

please, keep coming back!

File: sardana_albaem/ctrl/Albaem2OneDCtrl.py (hunk 1 above)
Comment (13bscsaamjad) [reply sardana_albaem/ctrl/Albaem2OneDCtrl.py]: i based it on the ct controller and took it from there, should i remove it?

File: sardana_albaem/ctrl/Albaem2OneDCtrl.py (hunk 3 above)
Comment (13bscsaamjad) [reply sardana_albaem/ctrl/Albaem2OneDCtrl.py:125]: this is sardana internal variable coming from the base class.

File: sardana_albaem/ctrl/Albaem2OneDCtrl.py (hunk 4 above)
Comment (13bscsaamjad) [reply sardana_albaem/ctrl/Albaem2OneDCtrl.py]: yeah could be, again basing it on the ct controller, thinking should not diverge much.

File: sardana_albaem/ctrl/Albaem2OneDCtrl.py (hunk 1 above)
Comment (meguiraun) [reply sardana_albaem/ctrl/Albaem2OneDCtrl.py]: If it is not used, better remove it

File: sardana_albaem/ctrl/Albaem2OneDCtrl.py (hunk 5 above)
Comment (13bscsaamjad) [reply sardana_albaem/ctrl/Albaem2OneDCtrl.py:126]: this is internal sardana variable coming from base class!

File: sardana_albaem/ctrl/Albaem2OneDCtrl.py (hunk 9 above)
Comment (13bscsaamjad) [reply sardana_albaem/ctrl/Albaem2OneDCtrl.py:286]: the data is read from ReadOne!

File: sardana_albaem/ctrl/Albaem2OneDCtrl.py (hunk 12 above)
Comment (13bscsaamjad) [reply sardana_albaem/ctrl/Albaem2OneDCtrl.py]: from the dashboard or some view to configure!

**Review by meguiraun — APPROVED** [review]

**Pull request merged**